src/modules/helpers/index.ts

```
@@ -445,6 +445,52 @@ export class HelpersModule {
     return array[index];
   }
 
+  /**
+   * Returns a weighted random element from the given array. Each element of the array should be an object with two keys `weight` and `value`.
+   *
+   * - Each `weight` key should be a number representing the probability of selecting the value, relative to the sum of the weights. Weights can be any positive float or integer.
+   * - Each `value` key should be the corresponding value.
+   *
+   * For example, if there are two values A and B, with weights 1 and 2 respectively, then the probability of picking A is 1/3 and the probability of picking B is 2/3.
+   *
+   * @template T The type of the entries to pick from.
+   * @param array Array to pick the value from.
+   *
+   * @example
+   * faker.helpers.weightedArrayElement([{ weight: 5, value: 'sunny' }, { weight: 4, value: 'rainy' }, { weight: 1, value: 'snowy' }]) // 'sunny', 50% of the time, 'rainy' 40% of the time, 'snowy' 10% of the time
+   *
+   * @since 8.0.0
+   */
+  weightedArrayElement<T>(
+    array: ReadonlyArray<{ weight: number; value: T }>
+  ): T {
+    if (array.length === 0) {
+      throw new FakerError(
+        'weightedArrayElement expects an array with at least one element'
+      );
+    }
+    if (!array.every((elt) => elt.weight > 0)) {
+      throw new FakerError(
+        'weightedArrayElement expects an array of { weight, value } objects where weight is a positive number'
+      );
+    }
+    const total = array.reduce((acc, { weight }) => acc + weight, 0);
+    const random = this.faker.number.float({
+      min: 0,
+      max: total,
+      precision: 1e-9,
+    });
+    let current = 0;
+    for (const { weight, value } of array) {
+      current += weight;
+      if (random < current) {
+        return value;
+      }
+    }
+    // In case of rounding errors, return the last element
+    return array[array.length - 1].value;
```

Can we somehow force this case in a test? I have no idea how sadly.

---

I could be wrong, but I think because of line 472 and 484-486 it is impossible to reach this code
So it is only more like a fallback to satisfy TypeScript, is it?

---

https://javascript.plainenglish.io/why-0-1-0-2-0-3-in-javascript-d7e218224a72

Adding floating point numbers in JS is weird

But it's practically not going to happen

---

Being certain it never happens requires a fairly good understanding of how JS floating point addition works, and the exact implementation of `faker.number.float` so i prefer to have this line as a fail-safe rather than risk returning undefined in certain pathological cases.

---

> Being certain it never happens requires a fairly good understanding of how JS floating point addition works, and the exact implementation of `faker.number.float` so i prefer to have this line as a fail-safe rather than risk returning undefined in certain pathological cases.

I totally get this and think it's a good way of implementing this. I was just wondering if we could somehow force this to actually prove your statement "look JS float is weird, that's why this is in place". 

I'm more afraid that someone in the future will have the same conclusion as @Shinigami92, remove the fall back, and will then run into a failing test. That's why I was asking for a possible test^^

---

IMO the likelihood of that is low, because we have an explicit comment why it is there.